content/en/blog/_posts/2020-05-14-my-exciting-journey-into-kubernetes-history.md

```diff
@@ -0,0 +1,656 @@
+
+# Building the Machine Learning Model
+
+Now we have an idea what the data set is about and we can start building a first
+machine learning model. Before actually building the model, we have to
+pre-process all the extracted release notes from the PRs, otherwise the model
+would not be able to understand our input.
+
+## Doing some first Natural Language Processing (NLP)
+
+In the beginning we have to define a vocabulary for which we want to train. I
+decided to choose the [TfidfVectorizer][60] from the Python scikit-learn machine
+learning library. This vectorizer is able to take our input texts and create a
+single huge vocabulary out of it. This is our so-called [bag-of-words][61],
+which has a chosen n-gram range of `(1, 2)` (unigrams and bigrams). Practically
+this means that we always use the first word and the next one as a single
+vocabulary entry (bigrams). We also use the single word as vocabulary entry
+(unigram). The TfidfVectorizer is able to skip words which occur multiple times
+(`max_df`) and requires a minimum amount (`min_df`) to add a word to the
+vocabulary. I decided not to change those values in the first place, just
+because I had the intuition that release notes are something unique to a
+project.
+
+Parameters like `min_df`, `max_df` and the n-gram range can be seen as some of
+our hyperparameters. Those parameters have to be optimized in a dedicated step
+after the machine learning model has been built. This step is called
+hyperparameter tuning and basically means that we train multiple times with
+different parameters and compare the accuracy of the model. Afterwards we choose
+the parameters with the best accuracy.
+
+[60]: https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html
+[61]: https://en.wikipedia.org/wiki/Bag-of-words_model
+
+During the training the vectorizer will produce a `data/features.json` which
+contains the whole vocabulary. This gives us a good understanding of how such a
+vocabulary may look like:
+
+```json
+[
+  …
+  "hostname",
+  "hostname address",
+  "hostname and",
+  "hostname as",
+  "hostname being",
+  "hostname bug",
+  …
+]
+```
+
+This produces round about 50.000 entries in the overall bag-of-words, which is
+pretty much. Previous analyses between different data sets showed that it is
+simply not necessary to take so many features into account. Some general data
+sets state that an overall vocabulary of 20.000 is enough and higher amounts do
+not influence the accuracy any more. To do so we can use the [SelectKBest][62]
+feature selector to strip down the vocabulary to only choose the top features.
+Anyway, I still decided to stick to the top 50.000 to not negatively influence
+the model accuracy. We have a relatively low amount of data (appr. 7.000
+samples) and a low number of words per sample (~15) which already made me wonder
+if we have enough data at all.
```

Comment:
nit: English locales usually use comma as the thousands separator.

Reply:
Fixed that as well.